File: frontend/src/services/project.ts
import api from './api';
import { Project } from '../types';

export const projectService = {
  async list(): Promise<Project[]> {
    const response = await api.get<Project[]>('/projects/');
    return response.data;
  },

  async get(id: number): Promise<Project> {
    const response = await api.get<Project>(`/projects/${id}`);
    return response.data;
  },

  async create(data: {
    name: string;
    description?: string;
    datahub_dataset_urn?: string;
    omaha_config?: string;
  }): Promise<Project> {
    const response = await api.post<Project>('/projects/', data);
    return response.data;
  },

  async update(
    id: number,
    data: {
      name?: string;
      description?: string;
      datahub_dataset_urn?: string;
      omaha_config?: string;
    }
  ): Promise<Project> {
    const response = await api.put<Project>(`/projects/${id}`, data);
    return response.data;
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/projects/${id}`);
  },
};

File: backend/app/schemas/project.py
"""
Project Pydantic schemas.
"""
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional


class ProjectBase(BaseModel):
    """Base project schema."""

    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    datahub_dataset_urn: Optional[str] = None


class ProjectCreate(ProjectBase):
    """Project creation schema."""

    omaha_config: Optional[str] = None  # YAML text


class ProjectUpdate(BaseModel):
    """Project update schema."""

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    datahub_dataset_urn: Optional[str] = None
    omaha_config: Optional[str] = None


class ProjectInDB(ProjectBase):
    """Project in database schema."""

    id: int
    owner_id: int
    omaha_config: Optional[str] = None
    project_metadata: dict = {}
    setup_stage: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class Project(ProjectInDB):
    """Project response schema."""

    pass


class ProjectWithOwner(Project):
    """Project with owner information."""

    owner_email: str
    owner_username: str
